rfcs/RangeSlider.md

```diff
@@ -0,0 +1,62 @@
+# RFC: Add RangeSlider option to existing Slider component
+
+## Summary
+
+We are proposing to do some changes to the props of the existing Slider to expand its functionality. With the new option users will be able to use it as a RangeSlider.
+
+## Background
+
+The Azure Portal team is gradually embracing React for the development of Azure Portal, however there is no React-based UI library internally for Portal. So we decided to use FluentUI for basic UI needs. Whenever we cannot find a good match for an old portal control from existing FluentUI components, we would like to collaborate and contribute back to FluentUI to fill the gap.
+
+## Problem
+
+There is a control called RangeSlider that is used by a couple of Azure extension teams. (https://ms.portal.azure.com/?Microsoft_Azure_Playground=true#blade/Microsoft_Azure_Playground/ControlsIndexBlade/RangeSlider_createSimpleRangeSlider_Playground), which is not supported by current FluentUI. The current Slider component from FluentUI has only one thumb that allows the users to drag and change value. Since this kind of control is generic and no logic is specific to Portal, we propose to add features to the existing FluentUI slider, to make it capable of becoming a RangeSlider. So that the portal teams using RangeSlider will not be blocked when converting to ReactView, and FluentUI Slider component could provide more features.
+
+## Detailed Proposal
+
+Current RangeSlider control in Azure Portal is more complex than the existing fluent Slider component, in that it has markers below the slider bar and textboxes on both sides that allows user input\. To make things simple, the proposal only aims at making the minimum function working. We will focus on adding a thumb to make it a RangeSlider, and not talk about the other functionalities.
```

Comment:
One general concern with adding major functionality like this to existing components is that it makes the code more complex (harder to maintain) and increases the bundle size. If there's a way to build this functionality as a separate component which reuses Slider and provides custom renderers and callbacks to handle the extra thumb, that would be preferable.

Reply:
And if Slider doesn't currently provide enough customization access to achieve what's needed for the RangeSlider, we could potentially add that.

Reply:
Are we sure that externalizing this is the right choice for range slider? 

We did discuss the idea of building a separate RangeSlider component already, but decided not to go to that extent right now. Instead we would make a incremental update to the existing slider and worry about the full featured slider (as is spec'd in figma) as part of slider convergence. 

Some thoughts:
- It seems like adding the ability to extend the number of thumbs and behavior of thumbs in the slider component would be more complex and perhaps nearly as much code as just adding range slider support? Perhaps I'm wrong here.
- Also, given we'll be converging slider soon and can figure out what the wholistic approach is here, it seems like taking the simple approach is a good path forward for now.

Reply:
I'm sure that the work for implementing the next gen Slider (as in figma) is way more than adding another thumb to the existing Slider. If you take a look at the sample code I posted in the RFC for RangeSlider, it is basically reusing all the existing slider logic, and only adds some custom logic to decide which thumb value to update in `onMouseDownOrTouchStart` and `updateValue`

Reply:
Per some early bundlesize analysis and further discussion of the timing of this change we decided to go with updating RangeSlider in place. 

Going forward with Slider convergence we probably should look at how Slider can be constructed such that it's easy to share the code to build both a Slider and a RangeSlider without having to expose all of the extension points on Slider's surface. That is work for another time and not part of this RFC.

Reply:
There is an important difference between Slider and RangeSlider as Slider can be inplenented using `<input type='range'>`. When implementing Slider for react-northstar we faced issues with accessibility when we wanted to implement it using styled divs - as we were getting off by 1 errors with voiceover and inability to control the value using VO keys. Having two separate components would allow us to minimize this impact only to the RangeSlider.

If we decide to go with styled divs, I think we should at least raise the issues against Chromium and/or Voiceover and get understanding when this can be fixed on their side.

Reply:
Yeah, the v7/v8 slider is styled divs/spans, so this is out of scope at least for v7/v8.

@jurokapsiar makes a good point that we should consider for converged Slider though. I think for converged slider it will make sense to have <Slider/> and <RangeSlider/>, but for v7/v8 the proposed approach makes sense as an incremental addition. 

Trying to get good code reuse across Slider and RangeSlider with a different DOM structure will be an interesting test of our compose pattern 😄